runtime/memprof.c

```diff
@@ -274,324 +378,293 @@ static value capture_callstack(int alloc_idx)
+  new_alloc_len = new_len * 2;
+  if (new_alloc_len < ea->min_alloc_len)
+    new_alloc_len = ea->min_alloc_len;
+  new_t = caml_stat_resize_noexc(ea->t, new_alloc_len * sizeof(struct tracked));
+  if (new_t == NULL) return 0;
+  ea->t = new_t;
+  ea->alloc_len = new_alloc_len;
   return 1;
 }
 
+#define Invalid_index (~(uintnat)0)
+
 Caml_inline uintnat new_tracked(uintnat n_samples, uintnat wosize,
                                 int is_unmarshalled, int is_young,
                                 value block, value user_data)
 {
   struct tracked *t;
-  trackst.len++;
-  if (!realloc_trackst()) {
-    trackst.len--;
+  if (!realloc_entries(&local->entries, 1))
     return Invalid_index;
-  }
-  t = &trackst.entries[trackst.len - 1];
+  local->entries.len++;
+  t = &local->entries.t[local->entries.len - 1];
   t->block = block;
   t->n_samples = n_samples;
   t->wosize = wosize;
   t->user_data = user_data;
-  t->idx_ptr = NULL;
+  t->running = NULL;
   t->alloc_young = is_young;
   t->unmarshalled = is_unmarshalled;
   t->promoted = 0;
   t->deallocated = 0;
-  t->cb_alloc_called = t->cb_promote_called = t->cb_dealloc_called = 0;
+  t->cb_promote_called = t->cb_dealloc_called = 0;
   t->deleted = 0;
-  t->callback_running = 0;
-  return trackst.len - 1;
+  return local->entries.len - 1;
 }
 
-static void mark_deleted(uintnat t_idx)
+static void mark_deleted(struct entry_array* ea, uintnat t_idx)
 {
-  struct tracked* t = &trackst.entries[t_idx];
+  struct tracked* t = &ea->t[t_idx];
   t->deleted = 1;
   t->user_data = Val_unit;
   t->block = Val_unit;
-  if (t_idx < trackst.delete) trackst.delete = t_idx;
-  CAMLassert(t->idx_ptr == NULL);
+  if (t_idx < ea->delete_idx) ea->delete_idx = t_idx;
 }
 
-/* The return value is an exception or [Val_unit] iff [*t_idx] is set to
-   [Invalid_index]. In this case, the entry is deleted.
-   Otherwise, the return value is a [Some(...)] block. */
-Caml_inline value run_callback_exn(uintnat *t_idx, value cb, value param) {
-  struct tracked* t = &trackst.entries[*t_idx];
+Caml_inline value run_callback_exn(
+  struct entry_array* ea, uintnat t_idx, value cb, value param)
+{
+  struct tracked* t = &ea->t[t_idx];
   value res;
-  CAMLassert(!t->callback_running && t->idx_ptr == NULL);
+  CAMLassert(t->running == NULL);
   CAMLassert(lambda > 0.);
 
-  callback_running = t->callback_running = 1;
-  t->idx_ptr = t_idx;
+  local->callback_status = ea == &entries_global ? t_idx : CB_LOCAL;
+  t->running = local;
+  t->user_data = Val_unit;      /* Release root. */
   res = caml_callback_exn(cb, param);
-  callback_running = 0;
-  /* The call above can modify [*t_idx] and thus invalidate [t]. */
-  if (*t_idx == Invalid_index) {
-    /* Make sure this entry has not been removed by [caml_memprof_set] */
-    return Val_unit;
+  if (local->callback_status == CB_STOPPED) {
+    /* Make sure this entry has not been removed by [caml_memprof_stop] */
+    local->callback_status = CB_IDLE;
```

Comment:
I don't understand why `local->callback_status` is changed here.

Reply:
Never mind, I do now! I hadn't read the comment on CB_STOPPED properly: a thread is CB_STOPPED while it is executing a callback during which Memprof was stopped, and CB_IDLE once that callback has finished.

Reply:
The documentation of `callback_status` says:

>   /* [callback_status] contains:
>     - CB_STOPPED if the current thread is running a callback, but
>       sampling has been stopped using [caml_memprof_stop];

Here, we go from the status "running a callback and stopped" to the status "not running a callback and stopped", hence `CB_STOPPED` is no longer applicable. This gets more coherent with the behavior of `caml_memprof_stop`, where we do:

```c
  if (ctx->callback_status != CB_IDLE) ctx->callback_status = CB_STOPPED;
```

I.e., we use `CB_STOPPED` only when a callback is currently running.

Moreover, not resetting `callback_status` would have the consequence that, at the next call to `caml_memprof_start`, `callback_status` would still contain `CB_STOPPED` (since `caml_memprof_start` does not touch this flag), which is clearly wrong. And no, you don't want `caml_memprof_start` to reset it, since that function could be called during a callback.